loader/loader.go

```diff
@@ -414,51 +415,97 @@ func formatInvalidKeyError(keyPrefix string, key interface{}) error {
 
 // LoadServices produces a ServiceConfig map from a compose file Dict
 // the servicesDict is not validated if directly used. Use Load() to enable validation
-func LoadServices(servicesDict map[string]interface{}, workingDir string, lookupEnv template.Mapping) ([]types.ServiceConfig, error) {
+func LoadServices(servicesDict map[string]interface{}, workingDir string, lookupEnv template.Mapping, opts *Options) ([]types.ServiceConfig, error) {
 	var services []types.ServiceConfig
 
-	for name, serviceDef := range servicesDict {
-		serviceConfig, err := LoadService(name, serviceDef.(map[string]interface{}), workingDir, lookupEnv)
+	for name := range servicesDict {
+		serviceConfig, err := loadServiceWithExtends(name, servicesDict, workingDir, lookupEnv, opts, 0)
 		if err != nil {
 			return nil, err
 		}
 
-		if serviceConfig.Extends != nil {
-			file := serviceConfig.Extends["file"]
-			service := serviceConfig.Extends["service"]
-			var source interface{}
-			if file == nil {
-				// extends a service from same file
-				source = servicesDict[*service]
-			} else {
-				if !filepath.IsAbs(*file) {
-					absolute := filepath.Join(workingDir, *file)
-					file = &absolute
-				}
-				bytes, err := ioutil.ReadFile(*file)
-				if err != nil {
-					return nil, err
-				}
-				source, err = ParseYAML(bytes)
+		services = append(services, *serviceConfig)
+	}
+
+	return services, nil
+}
+
+func loadServiceWithExtends(name string, servicesDict map[string]interface{}, workingDir string, lookupEnv template.Mapping, opts *Options, depth int) (*types.ServiceConfig, error) {
+	if depth > 100 {
```

Comment:
Would better track services we've already seen (like https://github.com/docker/compose/blob/master/compose/config/config.py#L635 does) to detect cycles. Could also track as a slice so we can report the cycle to end user for better diagnostic

Reply:
Ah yeah I considered that, but wasn't sure how to uniquely identify the service. Do you think I should modify the functions to take the filename as an arg as well?

Reply:
to mimic docker-compose this probably makes sense to refer to seen services as `filename:service` (see https://github.com/docker/compose/blob/master/compose/config/config.py#L585-L586). So passing filename to this func makes perfect sense to me

Reply:
I'm not sure how to get the filename of config file that's initially passed to `loadSections`. It doesn't look like that info is available anywhere.

~~I could imagine using the empty string to represent that initial file's filename, but then we wouldn't be able to detect cyclic dependencies to that file.~~

Actually, I think that would work. We wouldn't detect the cycle on the first pass, but after loading the service the second time, it'd have the filename set. So I'll plan to write it like that.

Reply:
@ndeloof I've added the improved cycle tracking